nncf/data/dataloader.py

```diff
@@ -0,0 +1,84 @@
+"""
```

Comment:
Please take a look on my proposal. I tried to implement statements that we discussed offline. Please feel free to comment.
```
class Sampler(ABC):

    def __iter__(self) -> Iterator[int]:
        pass



class InferenceDataLoader:

    def __init__(data_source: Iterable, transform_fn: Callable[[LabledDataItem], UnlabledDataItem], sampler:Optional[Sampler] = None):
        self._data_source = data_source

    def __iter__(self) -> Iterator[UnlabledDataItem]:
        pass



class ValidationDataLoader:

    def __init__(data_source: Iterable, sampler: Optional[Sampler] = None):
        self._data_source = data_source

    def __iter__(self) -> Iterator[LabledDataItem]:
        pass



class Dataset:

    def __init__(data_source: Iterable, transform_fn: Callable[[LabledDataItem], UnlabledDataItem]):
        self._data_source = data_source
        self._transform_fn = transform_fn

    @property
    def data_source(self) -> Iterable:
        return self._data_source
    

    def get_inference_dataloader(self, sampler: Optional[Sampler] = None) -> InferenceDataLoader:
        pass

    def get_validation_dataloader(self, sampler: Optional[Sampler] = None) -> ValidationDataLoader:
        pass



def val_func(model: ModelType, val_dataloader: Iterable[LabledDataItem]) -> float:
    pass


def transform_fn(data_item: LabledDataItem) -> UnlabledDataItem:
    pass


def quantize(model: ModelType,
             calibration_dataset: Dataset,
             preset: QuantizationPreset = QuantizationPreset.PERFORMANCE,
             target_device: TargetDevice = TargetDevice.ANY,
             subset_size: int = 300,
             fast_error_correction: bool = True,
             model_type: Optional[str] = None) -> ModelType:


def quantize_with_accuracy_control(model: ModelType,
                                   calibration_dataset: Dataset,
                                   validation_dataset: Dataset,
                                   validation_fn: Callable[[ModelType, Iterable[LabledDataItem]], float],
                                   max_drop: float = 0.01,
                                   preset: QuantizationPreset = QuantizationPreset.PERFORMANCE,
                                   target_device: TargetDevice = TargetDevice.ANY,
                                   subset_size: int = 300,
                                   fast_error_correction: bool = True,
                                   model_type: Optional[str] = None) -> ModelType:
```

Reply:
Several points from my side:
- I think it's better to use the following naming:
  ```python
  class ValidationDataLoader --> class LabeledDataLoader:

  class InferenceDataLoader --> class UnlabeledDataLoader:

  get_inference_dataloader(...) --> get_unlabeled_dataloader()

  get_validation_dataloader(...) --> get_labeled_dataloader()
  ```
  I think these names are more usual.
- Why do we need the `data_source` property in the `Dataset` class?

Reply:
> Several points from my side:
> 
>     * I think it's better to use the following naming:
>       ```python
>       class ValidationDataLoader --> class LabeledDataLoader:
>       
>       class InferenceDataLoader --> class UnlabeledDataLoader:
>       
>       get_inference_dataloader(...) --> get_unlabeled_dataloader()
>       
>       get_validation_dataloader(...) --> get_labeled_dataloader()
>       ```
>       
>       
>           
>             
>           
>       
>             
>           
>       
>           
>         
>       I think these names are more usual.
> 
>     * Why do we need the `data_source` property in the `Dataset` class?

Good question. Right now It is not needed.

Reply:
For user perspective:
How to run quantization:
1. Create nncf.Dataset. Let a developer has DataLoader or Dataset from ```torch.utils.data``` and called it is as `data_source` and the developer wrote `transform_fn`
```
dataset = nncf.Dataset(dataloader, transform_fn)
```
2. call the quantization method
```
q_model = nncf.quantize(model, calibration_dataset: dataset)
```
How to test transform_fn or/and validation_fn on user side: 
```
inference_dataloader = dataset.get_inference_dataloader()
try:
    model(next(iter(inference_dataloader)))
except:
    pass
```
```
 validation_dataloader = dataset.get_validation_dataloader()
try:
    validation_fn(model, validation_dataloader)
except:
    pass
```

Reply:
It was implemented.